hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/AzureBlobFileSystemStore.java

```diff
@@ -1418,6 +1428,73 @@ public void rename(final Path source, final Path destination,
     } while (shouldContinue);
   }
 
+  private void renameBlobDir(final Path source,
+      final Path destination,
+      final TracingContext tracingContext,
+      final ListBlobQueue listBlobQueue,
+      final BlobProperty blobPropOnSrc, final AbfsBlobLease srcDirBlobLease,
+      final Boolean isAtomicRename) throws AzureBlobFileSystemException {
+    BlobList blobList;
+    ListBlobConsumer listBlobConsumer = new ListBlobConsumer(listBlobQueue);
+    while(!listBlobConsumer.isCompleted()) {
+      blobList = listBlobConsumer.consume();
+      if(blobList == null) {
+        continue;
+      }
+      List<Future> futures = new ArrayList<>();
+      for (BlobProperty blobProperty : blobList.getBlobPropertyList()) {
+        futures.add(renameBlobExecutorService.submit(() -> {
+          try {
+            AbfsBlobLease blobLease = null;
+            if (isAtomicRename) {
+              /*
+               * Conditionally get a lease on the source blob to prevent other writers
+               * from changing it. This is used for correctness in HBase when log files
+               * are renamed. It generally should do no harm other than take a little
+               * more time for other rename scenarios. When the HBase master renames a
+               * log file folder, the lease locks out other writers.  This
+               * prevents a region server that the master thinks is dead, but is still
+               * alive, from committing additional updates.  This is different than
+               * when HBase runs on HDFS, where the region server recovers the lease
+               * on a log file, to gain exclusive access to it, before it splits it.
+               */
+              blobLease = new AbfsBlobLease(client,
+                  blobProperty.getPath().toUri().getPath(), tracingContext);
+            }
+            if(srcDirBlobLease != null) {
```

Comment:
Renewing only at this stage will lead to issues if copy takes longer leaving the src directory lease to expire. A background thread that will continously renew src folder lease is needed.

Reply:
Yes. Great point! In the next revision, have used the timer to renew the lease.